examples/java/src/main/java/org/apache/beam/examples/complete/game/UserScore.java

```diff
@@ -108,8 +112,39 @@ public String getKey(String keyname) {
       }
     }
 
-    public Long getTimestamp() {
-      return this.timestamp;
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
```

Comment:
There is a subtle thing here. If you want to insist that the class is exactly this class, then you should use `o.getClass() == GameActionInfo.class`. That matches your use of field references below. Otherwise, you should change each `this.team` into `this.getTeam()`, and so on for score, user, etc. Basically, embrace inheritance or prevent inheritance. FWIW inheritance and equivalence relations are opposed anyhow.

Reply:
Great comment. I liked the way you laid out the options, it was a nice way for me to learn about the subtleties. Thanks! I ended up choosing it `o.getClass() == GameActionInfo.class` option so it is explicit we're checking this specific class.

Reply:
nevermind, I had to revert it, FindBugs reported it:

```EQ_GETCLASS_AND_CLASS_CONSTANT: equals method fails for subtypes
This class has an equals method that will be broken if it is inherited by subclasses. It compares a class literal with the class of the argument (e.g., in class Foo it might check if Foo.class == o.getClass()). It is better to check if this.getClass() == o.getClass().``` 

So I changed it from the fields to the getter methods.

Reply:
Findbugs and I have a mild feud over this one, as I care more about the usual properties of equality than the usual properties of inheritance. One is much older and more clearly timeless than the other... :-)

Your change is g2g though. This is a corner case with no right answer, so making findbugs happy by calling a getter is just fine with me.

Reply:
hah, thanks!